// @oasisdigital/angular-material-search-select/esm5/base.js
// Base class with which to create an observable auto complete. This doesn't
// have any connection to how the result is rendered in the Dom, with what
// component set, and so on.
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { FormControl } from '@angular/forms';
import { BehaviorSubject, of, EMPTY, timer, combineLatest, Subject } from 'rxjs';
import { switchMap, startWith, catchError, map, filter, debounce, take, refCount, withLatestFrom, publishReplay, distinctUntilChanged } from 'rxjs/operators';
var SearchSelectBase = /** @class */ (function () {
    function SearchSelectBase() {
        var _this = this;
        this.debounceTime = 75;
        // API provided to client component (for use in its template)
        this.searchControl = new FormControl();
        // ---------------------------------------------------------------------
        // Internals
        // tslint:disable:member-ordering
        this.incomingValues = new Subject();
        this.incomingDataSources = new BehaviorSubject({
            displayValue: of,
            search: function () { return of([]); }
        });
        this.onChange = function (_) { };
        this.onTouched = function () { };
        var searches = this.searchControl.valueChanges.pipe(startWith(this.searchControl.value), distinctUntilChanged(), debounce(function (srch) {
            // Typing into input sends strings.
            if (typeof srch === 'string') {
                return timer(_this.debounceTime);
            }
            return EMPTY; // immediate - no debounce for choosing from the list
        }));
        var options = combineLatest(searches, this.incomingDataSources.pipe(filter(function (ds) { return !!ds; }))).pipe(switchMap(function (_a) {
            var _b = tslib_1.__read(_a, 2), srch = _b[0], ds = _b[1];
            // Initial value is sometimes null.
            if (srch === null) {
                srch = '';
            }
            // Typing into input sends strings.
            if (typeof srch === 'string') {
                var search_1 = srch;
                return ds.search(srch).pipe(map(function (list) { return ({ search: search_1, list: list }); }), catchError(function (errorMessage) { return of({ search: search_1, errorMessage: errorMessage }); }), startWith({ search: search_1 }));
            }
            // Selecting from Material Option List sends an object, so there is
            // no need to call function to search for it.
            var entry = srch;
            return of({
                search: srch.display,
                list: [tslib_1.__assign({}, entry)]
            });
        }), publishReplay(1), refCount());
        function matcher(search, entry) {
            return entry.display === search;
        }
        this.selectedValue = options.pipe(filter(function (result) { return !!result.list; }), withLatestFrom(this.incomingDataSources), map(function (_a) {
            var _b = tslib_1.__read(_a, 2), result = _b[0], ds = _b[1];
            var list = result.list || []; // appease TS
            var matchFn = ds.match || matcher;
            var entry = list.find(function (option) { return matchFn(result.search, option); });
            return entry && entry.value || null;
        }), distinctUntilChanged());
        this.loading = options.pipe(map(function (o) { return !o.list && !o.errorMessage; }));
        this.list = options.pipe(map(function (o) { return o.list; }));
        this.empty = options.pipe(map(function (o) { return o.list ? o.list.length === 0 : false; }));
        this.errorMessage = options.pipe(map(function (o) { return o.errorMessage; }));
        // a value was provided by the form; request the full entry
        this.incomingDataSourcesSub = this.incomingValues.pipe(withLatestFrom(this.incomingDataSources), switchMap(function (_a) {
            var _b = tslib_1.__read(_a, 2), value = _b[0], ds = _b[1];
            return ds.displayValue(value);
        }))
            .subscribe(function (value) { return _this.searchControl.setValue(value); });
    }
    Object.defineProperty(SearchSelectBase.prototype, "dataSource", {
        set: function (ds) { this.incomingDataSources.next(ds); },
        enumerable: true,
        configurable: true
    });
    SearchSelectBase.prototype.focus = function () {
        // While focused, user selection will be propagated to the form.
        this.selectedValueSub = this.selectedValue.subscribe(this.checkAndPropagate.bind(this));
    };
    SearchSelectBase.prototype.blur = function () {
        this.onTouched();
        // Now that we've lost focus, stop propagating changes.
        if (this.selectedValueSub) {
            this.selectedValueSub.unsubscribe();
        }
        // However, it's possible the user has just typed some text that will be
        // confirmed (by the application provided function) as a match,
        // asynchronously. We can't force the system to wait for that to happen, we
        // are losing focus right now. But we can subscribe to pick up that one last
        // change and propagate it if/when it arrives.
        this.selectedValue
            .pipe(take(1))
            .subscribe(this.checkAndPropagate.bind(this));
        // However, this code raises an important question about valid behavior of a
        // Angular form control. Is it acceptable for a form control to
        // asynchronously provide a new value when it no longer has focus?
    };
    SearchSelectBase.prototype.displayWith = function (value) {
        return value ? value.display : '';
    };
    // Implement ControlValueAccessor
    SearchSelectBase.prototype.writeValue = function (obj) {
        // Angular sometimes writes a value that didn't really change.
        if (obj !== this.outsideValue) {
            this.outsideValue = obj;
            this.incomingValues.next(obj);
        }
    };
    SearchSelectBase.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    SearchSelectBase.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    SearchSelectBase.prototype.setDisabledState = function (isDisabled) {
        if (isDisabled) {
            this.searchControl.disable();
        }
        else {
            this.searchControl.enable();
        }
    };
    SearchSelectBase.prototype.ngOnDestroy = function () {
        if (this.incomingDataSourcesSub) {
            this.incomingDataSourcesSub.unsubscribe();
        }
    };
    SearchSelectBase.prototype.checkAndPropagate = function (value) {
        // Only send a change if there really is one.
        if (value !== this.outsideValue) {
            this.outsideValue = value;
            this.onChange(value);
        }
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], SearchSelectBase.prototype, "debounceTime", void 0);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], SearchSelectBase.prototype, "dataSource", null);
    return SearchSelectBase;
}());
export { SearchSelectBase };
