// @oasisdigital/angular-material-search-select/esm2015/base.js
// Base class with which to create an observable auto complete. This doesn't
// have any connection to how the result is rendered in the Dom, with what
// component set, and so on.
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { FormControl } from '@angular/forms';
import { BehaviorSubject, of, EMPTY, timer, combineLatest, Subject } from 'rxjs';
import { switchMap, startWith, catchError, map, filter, debounce, take, refCount, withLatestFrom, publishReplay, distinctUntilChanged } from 'rxjs/operators';
export class SearchSelectBase {
    constructor() {
        this.debounceTime = 75;
        // API provided to client component (for use in its template)
        this.searchControl = new FormControl();
        // ---------------------------------------------------------------------
        // Internals
        // tslint:disable:member-ordering
        this.incomingValues = new Subject();
        this.incomingDataSources = new BehaviorSubject({
            displayValue: of,
            search: () => of([])
        });
        this.onChange = (_) => { };
        this.onTouched = () => { };
        const searches = this.searchControl.valueChanges.pipe(startWith(this.searchControl.value), distinctUntilChanged(), debounce(srch => {
            // Typing into input sends strings.
            if (typeof srch === 'string') {
                return timer(this.debounceTime);
            }
            return EMPTY; // immediate - no debounce for choosing from the list
        }));
        const options = combineLatest(searches, this.incomingDataSources.pipe(filter(ds => !!ds))).pipe(switchMap(([srch, ds]) => {
            // Initial value is sometimes null.
            if (srch === null) {
                srch = '';
            }
            // Typing into input sends strings.
            if (typeof srch === 'string') {
                const search = srch;
                return ds.search(srch).pipe(map(list => ({ search, list })), catchError(errorMessage => of({ search, errorMessage })), startWith({ search }));
            }
            // Selecting from Material Option List sends an object, so there is
            // no need to call function to search for it.
            const entry = srch;
            return of({
                search: srch.display,
                list: [Object.assign({}, entry)]
            });
        }), publishReplay(1), refCount());
        function matcher(search, entry) {
            return entry.display === search;
        }
        this.selectedValue = options.pipe(filter(result => !!result.list), withLatestFrom(this.incomingDataSources), map(([result, ds]) => {
            const list = result.list || []; // appease TS
            const matchFn = ds.match || matcher;
            const entry = list.find(option => matchFn(result.search, option));
            return entry && entry.value || null;
        }), distinctUntilChanged());
        this.loading = options.pipe(map(o => !o.list && !o.errorMessage));
        this.list = options.pipe(map(o => o.list));
        this.empty = options.pipe(map(o => o.list ? o.list.length === 0 : false));
        this.errorMessage = options.pipe(map(o => o.errorMessage));
        // a value was provided by the form; request the full entry
        this.incomingDataSourcesSub = this.incomingValues.pipe(withLatestFrom(this.incomingDataSources), switchMap(([value, ds]) => ds.displayValue(value)))
            .subscribe(value => this.searchControl.setValue(value));
    }
    set dataSource(ds) { this.incomingDataSources.next(ds); }
    focus() {
        // While focused, user selection will be propagated to the form.
        this.selectedValueSub = this.selectedValue.subscribe(this.checkAndPropagate.bind(this));
    }
    blur() {
        this.onTouched();
        // Now that we've lost focus, stop propagating changes.
        if (this.selectedValueSub) {
            this.selectedValueSub.unsubscribe();
        }
        // However, it's possible the user has just typed some text that will be
        // confirmed (by the application provided function) as a match,
        // asynchronously. We can't force the system to wait for that to happen, we
        // are losing focus right now. But we can subscribe to pick up that one last
        // change and propagate it if/when it arrives.
        this.selectedValue
            .pipe(take(1))
            .subscribe(this.checkAndPropagate.bind(this));
        // However, this code raises an important question about valid behavior of a
        // Angular form control. Is it acceptable for a form control to
        // asynchronously provide a new value when it no longer has focus?
    }
    displayWith(value) {
        return value ? value.display : '';
    }
    // Implement ControlValueAccessor
    writeValue(obj) {
        // Angular sometimes writes a value that didn't really change.
        if (obj !== this.outsideValue) {
            this.outsideValue = obj;
            this.incomingValues.next(obj);
        }
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        if (isDisabled) {
            this.searchControl.disable();
        }
        else {
            this.searchControl.enable();
        }
    }
    ngOnDestroy() {
        if (this.incomingDataSourcesSub) {
            this.incomingDataSourcesSub.unsubscribe();
        }
    }
    checkAndPropagate(value) {
        // Only send a change if there really is one.
        if (value !== this.outsideValue) {
            this.outsideValue = value;
            this.onChange(value);
        }
    }
}
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object)
], SearchSelectBase.prototype, "debounceTime", void 0);
tslib_1.__decorate([
    Input(),
    tslib_1.__metadata("design:type", Object),
    tslib_1.__metadata("design:paramtypes", [Object])
], SearchSelectBase.prototype, "dataSource", null);
